packages/vuetify/src/composables/validation.ts

```diff
@@ -0,0 +1,80 @@
+// Utilities
+import { propsFactory, wrapInPromise } from '@/util'
+import { ref, unref, watch } from 'vue'
+
+// Types
+import type { MaybeRef } from '@/util/helpers'
+import type { PropType } from 'vue'
+
+export type ValidationRule = string | ((value: any) => string | boolean)
+export type ValidationRules = ValidationRule[]
+export type ValidationResult = string | boolean | Promise<string | boolean>
+
+export interface ValidationProps {
+  errorCount: string | number
+  rules: ValidationRules
+}
+
+export const makeValidationProps = propsFactory({
+  errorCount: {
+    type: [Number, String],
+    default: 1,
+  },
+  rules: {
+    type: Array as PropType<ValidationRules>,
+    default: () => ([]),
+  },
+})
+
+export function useValidation (props: ValidationProps, value: MaybeRef<any>) {
+  const errorMessages = ref<string[]>([])
+  const isPristine = ref(true)
+  const isValid = ref<null | boolean>(null)
+  const isValidating = ref(false)
+  const errorCount = ref(0)
+  const unwrefed = unref(value)
+
+  watch(isValid, () => isPristine.value = false)
+
+  async function validate () {
+    await reset()
+
+    isValidating.value = true
+    for (const rule of props.rules) {
+      const handler = typeof rule === 'function' ? rule : () => rule
+      const valid = await wrapInPromise<ValidationResult>(handler(unwrefed.value))
```

Comment:
See comment above about why unref is taking place. But assuming an unref, then it's the raw value, so `.value` should not be used.

Reply:
This was done under the pretense that we would accept anything, hence MaybeRef.

Reply:
Why should the composable accept something that is not a ref? I get that the actual input value could be anything, but that's `Ref<any / unknown>`.

Reply:
Not disagreeing, I hadn't removed it yet.